src/engine/QueryPlanner.cpp

```diff
@@ -952,6 +954,49 @@ ParsedQuery::GraphPattern QueryPlanner::seedFromInverse(
   return seedFromPropertyPath(right, path.children_[0], left);
 }
 
+// _____________________________________________________________________________
+ParsedQuery::GraphPattern QueryPlanner::seedFromNegated(
+    const TripleComponent& left, const PropertyPath& path,
+    const TripleComponent& right) {
+  AD_CORRECTNESS_CHECK(!path.children_.empty());
+  Variable forwardVar = generateUniqueVarName();
+  Variable inverseVar = generateUniqueVarName();
+  ParsedQuery::GraphPattern forwardPattern =
+      seedFromIri(left, PropertyPath::fromVariable(forwardVar), right);
+  ParsedQuery::GraphPattern inversePattern =
+      seedFromIri(right, PropertyPath::fromVariable(inverseVar), left);
+  auto makeFilter = [](const PropertyPath& unwrapped,
+                       const Variable& variable) {
+    using namespace sparqlExpression;
+    return SparqlFilter{
+        {std::make_shared<NotEqualExpression>(NotEqualExpression::Children{
+             std::make_unique<IriExpression>(
+                 TripleComponent::Iri::fromStringRepresentation(
+                     unwrapped.iri_)),
+             std::make_unique<VariableExpression>(variable)}),
+         absl::StrCat(unwrapped.iri_, " != ", variable.name())}};
+  };
+  using Operation = PropertyPath::Operation;
+  for (const auto& child : path.children_) {
+    if (child.operation_ == Operation::INVERSE) {
+      const auto& unwrapped = child.children_.at(0);
+      AD_CORRECTNESS_CHECK(unwrapped.operation_ == Operation::IRI);
+      inversePattern._filters.emplace_back(makeFilter(unwrapped, inverseVar));
+    } else {
+      AD_CORRECTNESS_CHECK(child.operation_ == Operation::IRI);
+      forwardPattern._filters.emplace_back(makeFilter(child, forwardVar));
+    }
+  }
+  if (inversePattern._filters.empty()) {
+    return forwardPattern;
+  }
+  if (forwardPattern._filters.empty()) {
+    return inversePattern;
+  }
+  return uniteGraphPatterns(
```

Comment:
What are the semantics (in your code + the standard) for paths like
`?x !(<a> | <a>) ?y` , I think your code currently doesn't duplicate the triples, but also probably no one cares.

Reply:
The current behaviour is that if you dare to repeat the same iri twice it's on you to take the performance toll.